### PR TITLE
[feat]: [OIP-454]: Save API Call request Body for CVNGLog

### PR DESCRIPTION
**Pull request by n00bitax**
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/48484)

### CODE REVIEW

File: srm-service/modules/cv-nextgen-service/contracts/commons/src/test/java/io/harness/cvng/beans/cvnglog/ApiCallLogDTOTest.java
Comment (dhruv-harness) [inline srm-service/modules/cv-nextgen-service/contracts/commons/src/test/java/io/harness/cvng/beans/cvnglog/ApiCallLogDTOTest.java:1]: Let us add test for adding response body in the log for CVNGDataCollectionDelegateServiceImpl.java

**Comment by n00bitax** [comment]
trigger uts

File: 957-cg-beans/src/main/java/io/harness/cvng/beans/cvnglog/ApiCallLogUtils.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+import lombok.extern.slf4j.Slf4j;
+import okhttp3.Request;
+import okio.Buffer;
+import org.apache.commons.lang.StringEscapeUtils;
+
+@OwnedBy(CV)
+@UtilityClass
+@Slf4j
+public class ApiCallLogUtils {
+  static String bodyToString(final Request request) {
```
[the remark this answers is not in the document]
Comment (n00bitax) [reply 957-cg-beans/src/main/java/io/harness/cvng/beans/cvnglog/ApiCallLogUtils.java:27]: static is required but yes it should be public, I have changed

File: 260-delegate/src/main/java/io/harness/perpetualtask/datacollection/ThirdPartyCallHandler.java [hunk 2]
```diff
@@ -56,11 +61,24 @@ public void accept(CallDetails callDetails) {
...
-
+    String method = callDetails.getRequest().request().method();
+    cvngLogDTO.addFieldToRequest(
+        ApiCallLogDTOField.builder().type(ApiCallLogDTO.FieldType.TEXT).name(REQUEST_METHOD).value(method).build());
+    Set<String> headerNames = callDetails.getRequest().request().headers().names();
+    if (headerNames.size() > 0) {
+      cvngLogDTO.addFieldToRequest(ApiCallLogDTOField.builder()
+                                       .type(ApiCallLogDTO.FieldType.TEXT)
+                                       .name(REQUEST_HEADERS)
+                                       .value(JsonUtils.asJson(headerNames))
```
[the remark this answers is not in the document]
Comment (n00bitax) [reply 260-delegate/src/main/java/io/harness/perpetualtask/datacollection/ThirdPartyCallHandler.java:76]: For now we are adding header names, not the values.

File: 957-cg-beans/src/main/java/io/harness/cvng/beans/cvnglog/ApiCallLogUtils.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+      final Request copy = request.newBuilder().build();
+      final Buffer buffer = new Buffer();
+      Objects.requireNonNull(copy.body()).writeTo(buffer);
+      return StringEscapeUtils.unescapeJava(buffer.readUtf8());
+    } catch (final IOException e) {
+      return "cannot parse the byteArray Request Body";
+    }
+  }
+
+  static ApiCallLogDTO.FieldType getFieldType(Request request) {
```
[the remark this answers is not in the document]
Comment (n00bitax) [reply 957-cg-beans/src/main/java/io/harness/cvng/beans/cvnglog/ApiCallLogUtils.java]: changed to mapRequestBodyContentTypeToFieldType

File: 957-cg-beans/src/main/java/io/harness/cvng/beans/cvnglog/ApiCallLogDTO.java [hunk 4]
```diff
@@ -50,25 +58,37 @@ public void addFieldToRequest(ApiCallLogDTOField field) {
     requests.add(field);
   }
 
+  public void addCallDetailsFieldToRequest(Request request) {
```
[the remark this answers is not in the document]
Comment (n00bitax) [reply 957-cg-beans/src/main/java/io/harness/cvng/beans/cvnglog/ApiCallLogDTO.java]: addCallDetailsBodyFieldToRequest

File: srm-service/modules/cv-nextgen-service/contracts/commons/src/test/java/io/harness/cvng/beans/cvnglog/ApiCallLogDTOTest.java
Comment (n00bitax) [reply srm-service/modules/cv-nextgen-service/contracts/commons/src/test/java/io/harness/cvng/beans/cvnglog/ApiCallLogDTOTest.java:1]: added ThirdPartyCallHandlerTest which covers e2e testing. Will add it that place later, I am adding a TODO

File: 260-delegate/src/main/java/io/harness/perpetualtask/datacollection/ThirdPartyCallHandler.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (dhruv-harness) [reply 260-delegate/src/main/java/io/harness/perpetualtask/datacollection/ThirdPartyCallHandler.java:76]: oh my bad

File: 957-cg-beans/src/main/java/io/harness/cvng/beans/cvnglog/ApiCallLogUtils.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dhruv-harness) [reply 957-cg-beans/src/main/java/io/harness/cvng/beans/cvnglog/ApiCallLogUtils.java]: can you please confirm once https://projectlombok.org/features/experimental/UtilityClass

**Comment by dhruv-harness** [comment]
trigger uts

**Review by dhruv-harness — APPROVED** [review]

**Comment by n00bitax** [comment]
trigger uts

**Comment by n00bitax** [comment]
trigger uts

**Comment by n00bitax** [comment]
trigger ut6

**Comment by n00bitax** [comment]
trigger ut1

**Comment by n00bitax** [comment]
trigger ut3

**Comment by n00bitax** [comment]
trigger ut1

**Comment by n00bitax** [comment]
trigger ut3

**Comment by harshjain12** [comment]
trigger uts

**Review by AbhijithHarness — APPROVED** [review]

**Comment by n00bitax** [comment]
trigger uts

**Pull request merged**